File: CCLib/src/cc/native/SystemTerminal.lua
natives["cc.terminal.SystemTerminal"] = natives["cc.terminal.SystemTerminal"] or {}

natives["cc.terminal.SystemTerminal"]["write(C)V"] = function(this, c)
	term.write(string.char(c))
end

natives["cc.terminal.SystemTerminal"]["write(Ljava/lang/String;)V"] = function(this, text)
	term.write(toLString(text))
end

natives["cc.terminal.SystemTerminal"]["clearLine()V"] = function(this)
	term.clearLine()
end

natives["cc.terminal.SystemTerminal"]["clear()V"] = function(this)
	term.clear();
end

natives["cc.terminal.SystemTerminal"]["getCursorX()I"] = function(this)
	local x, y = term.getCursorPos()
	return x - 1
end

natives["cc.terminal.SystemTerminal"]["getCursorY()I"] = function(this)
	local x, y = term.getCursorPos()
	return y - 1
end

natives["cc.terminal.SystemTerminal"]["setCursor(II)V"] = function(this, x, y)
	term.setCursorPos(x + 1, y + 1)
end

natives["cc.terminal.SystemTerminal"]["isColor()Z"] = function(this)
	return term.isColor() and 1 or 0
end

natives["cc.terminal.SystemTerminal"]["width()I"] = function(this)
	local w, h = term.getSize()
	return w
end

natives["cc.terminal.SystemTerminal"]["height()I"] = function(this)
	local w, h = term.getSize()
	return h
end

natives["cc.terminal.SystemTerminal"]["scroll(I)V"] = function(this, n)
	term.scroll(n)
end

natives["cc.terminal.SystemTerminal"]["setTextColor(Lcc/terminal/Color;)V"] = function(this, c)
	term.setTextColor(getObjectField(c, "intValue"))
end

natives["cc.terminal.SystemTerminal"]["setBackgroundColor(Lcc/terminal/Color;)V"] = function(this, c)
	term.setBackgroundColor(getObjectField(c, "intValue"))
end

File: CCLib/src/cc/terminal/SystemTerminal.java
package cc.terminal;

public class SystemTerminal implements Terminal {
	static {
		System.load("cc/native/SystemTerminal.lua");
	}
	
	public void write(char[] c) {
		for (char ch : c) {
			write(ch);
		}
	}

	@Override public native void write(char c);
	@Override public native void write(String text);

	@Override public native void clearLine();
	@Override public native void clear();

	@Override public native int getCursorX();
	@Override public native int getCursorY();
	@Override public native void setCursor(int x, int y);

	@Override public native boolean isColor();

	@Override public native int width();
	@Override public native int height();

	@Override public native void scroll(int n);

	@Override public native void setTextColor(Color c);
	@Override public native void setBackgroundColor(Color c);
	
	public void setColor(Color textColor, Color backgroundColor) {
		setTextColor(textColor);
		setBackgroundColor(backgroundColor);
	}

	public int nextLine() {
		return nextLine(getCursorX());
	}

	public int nextLine(int x) {
		if (getCursorY() == height() - 1) {
			scroll(1);
			setCursor(x, getCursorY());
		} else {
			setCursor(x, getCursorY() + 1);
		}
		return getCursorY();
	}
	
}